ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java

```diff
@@ -606,45 +605,52 @@ public void onCompletion(final BlobInfo routerResult, Exception routerException)
-                      response = new ByteBufferReadableStreamChannel(ByteBuffer.wrap(routerResult.getUserMetadata()));
-                    } else {
-                      setUserMetadataHeaders(userMetadata, restResponseChannel);
-                      restResponseChannel.setHeader(RestUtils.Headers.CONTENT_LENGTH, 0);
-                      response = new ByteBufferReadableStreamChannel(AmbryBlobStorageService.EMPTY_BUFFER);
+          securityService
+              .processResponse(restRequest, restResponseChannel, routerResult.getBlobInfo(), new Callback<Void>() {
+                @Override
+                public void onCompletion(Void securityResult, Exception securityException) {
+                  securityCallbackTracker.markOperationEnd();
+                  ReadableStreamChannel response = null;
+                  boolean blobNotModified = restResponseChannel.getStatus() == ResponseStatus.NotModified;
+                  try {
+                    if (securityException == null) {
+                      if (subResource != null) {
+                        BlobInfo blobInfo = routerResult.getBlobInfo();
+                        Map<String, String> userMetadata = RestUtils.buildUserMetadata(blobInfo.getUserMetadata());
+                        if (userMetadata == null) {
+                          restResponseChannel.setHeader(RestUtils.Headers.CONTENT_TYPE, "application/octet-stream");
+                          restResponseChannel
+                              .setHeader(RestUtils.Headers.CONTENT_LENGTH, blobInfo.getUserMetadata().length);
+                          response = new ByteBufferReadableStreamChannel(ByteBuffer.wrap(blobInfo.getUserMetadata()));
+                        } else {
+                          setUserMetadataHeaders(userMetadata, restResponseChannel);
+                          restResponseChannel.setHeader(RestUtils.Headers.CONTENT_LENGTH, 0);
+                          response = new ByteBufferReadableStreamChannel(AmbryBlobStorageService.EMPTY_BUFFER);
+                        }
+                      } else if (!blobNotModified) {
+                        response = routerResult.getBlobDataChannel();
+                      }
+                    }
+                  } catch (Exception e) {
+                    frontendMetrics.getSecurityResponseCallbackProcessingError.inc();
+                    securityException = e;
+                  } finally {
+                    if (response != null || securityException != null || blobNotModified) {
+                      submitResponse(restRequest, restResponseChannel, response, securityException);
+                    }
+                    if (securityException != null && routerResult.getBlobDataChannel() != null) {
+                      try {
+                        routerResult.getBlobDataChannel().close();
+                      } catch (IOException ignored){
+                      }
```

Comment:
Actually, you might need to retain this code but for a different condition. Exceptions are fine but what about the blob not modified case? Don't you need to close the `ReadableStreamChannel` then?

I just thought about this and this opens up a few more questions. The other reason for the two different calls was probably for the not-modified use case. If the blob was deemed not modified, we would not fetch it.

I think we will require one more change irrespective of all of this (this need not be done right now but please open an issue). The router should fetch only one data chunk (in case of large blob) until the channel starts to be read. Once the channel is read, it can fill 4 chunks at a time. Otherwise, the not modified case would fetch 4 * 4 = 16 MB unnecessarily.